src/instructlab/model/download.py

```diff
@@ -12,6 +16,236 @@
+        )
+
+        if self.hf_token == "" and "instructlab" not in self.repository:
+            raise ValueError(
+                """HF_TOKEN var needs to be set in your environment to download HF Model.
+                Alternatively, the token can be passed with --hf-token flag.
+                The HF Token is used to authenticate your identity to the Hugging Face Hub."""
+            )
+
+        try:
+            if self.ctx.obj is not None:
+                hf_logging.set_verbosity(self.ctx.obj.config.general.log_level.upper())
+            files = list_repo_files(repo_id=self.repository, token=self.hf_token)
+            if any(".safetensors" in string for string in files):
+                self.download_safetensors()
+            else:
+                self.download_gguf()
+
+        except Exception as exc:
+            click.secho(
+                f"Downloading model failed with the following Hugging Face Hub error: {exc}",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+
+    def download_gguf(self) -> None:
+        try:
+            hf_hub_download(
+                token=self.hf_token,
+                repo_id=self.repository,
+                revision=self.release,
+                filename=self.filename,
+                local_dir=self.download_dest,
+            )
+
+        except Exception as exc:
+            click.secho(
+                f"Downloading GGUF model failed with the following HuggingFace Hub error: {exc}",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+
+    def download_safetensors(self) -> None:
+        try:
+            if not os.path.exists(os.path.join(self.download_dest, self.repository)):
+                os.makedirs(
+                    name=os.path.join(self.download_dest, self.repository),
+                    exist_ok=True,
+                )
+            snapshot_download(
+                token=self.hf_token,
+                repo_id=self.repository,
+                revision=self.release,
+                local_dir=os.path.join(self.download_dest, self.repository),
+            )
+        except Exception as exc:
+            click.secho(
+                f"Downloading safetensors model failed with the following HuggingFace Hub error: {exc}",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+
+
+class OCIDownloader(Downloader):
+    """Class to handle downloading safetensors models from OCI Registries"""
+
+    def __init__(
+        self,
+        repository: str,
+        release: str,
+        download_dest: str,
+    ) -> None:
+        super().__init__(
+            repository=repository, release=release, download_dest=download_dest
+        )
+        self.repository = repository
+        self.release = release
+        self.download_dest = download_dest
+
+    def _build_oci_model_file_map(self, oci_model_path: str) -> dict:
```

Comment:
About this code, how do we ensure that the model does not change and this code doesn't break? What OCI releases a new spec/version and we lose compatibility?

Is skopeo, capable of doing this? Can skopeo give us the information we are looking for by reading the directory/image?

Reply:
I spoke with @n1hility about that. We need to write down which version of the OCI spec was used to write this logic.

Reply:
@leseb what would the best place to document that be? should it be in this PR?

Reply:
Doc string in code + commit message. Add as much details as you can, thanks 🙏🏻

Reply:
added a note with version in both places